### ISSUE TITLE
Truncate or fold file paths if out of bounds

### ISSUE THREAD

**Issue by vuki656** [nvim-telescope/telescope.nvim#895]
**Is your feature request related to a problem? Please describe.**
Yes, in a large project, there are often deeply nested files leading to file paths going out of bounds and not being visible:

![image](https://user-images.githubusercontent.com/8104698/120899753-8ca12400-c631-11eb-89ff-cc0ee5c9349b.png)

Here a part of the file name is cut off, and if there was another file 1 level deeper, its name wouldn't be visible at all. This makes it difficult to spot what you are looking for. 

**Describe the solution you'd like**
Not 100% sure but either to truncate the file path if too long, or fold it.

Another potential solution to this would be to display only file names in the list and the path next to them, similarly to how IntelliJ does it.

![image](https://user-images.githubusercontent.com/8104698/120899967-9ecf9200-c632-11eb-829b-0c68e6e14d8d.png)

**Comment by Conni2461** [comment]
`Telescope find_files shorten_path=true` but note there will be a overhaul of how the filepath display options work: https://github.com/nvim-telescope/telescope.nvim/pull/839 this will also be cool https://github.com/nvim-telescope/telescope.nvim/pull/886

For making it dynamic there is a little bit missing because currently we don't know the result window width, when we calculate the display string, but i have an idea how we could do this.

**Comment by vuki656** [comment]
> `Telescope find_files shorten_path=true` but note there will be a overhaul of how the filepath display options work: #839 this will also be cool #886
> 
> For making it dynamic there is a little bit missing because currently we don't know the result window width, when we calculate the display string, but i have an idea how we could do this.

That helps. Although I'm still looking forward as to what they will come up with in #839.

Also, not sure if I'm doing something wrong, but when I launch `Telescope find_files shorten_path=true` it works as expected, but If I set it in the config like so it doesn't work. Am I doing something wrong? 

```
require('telescope').setup{
    shorten_path=true
}
```

**Comment by Conni2461** [comment]
Yeah currently it doesn't work in the `setup` table (even tho the readme says so, no idea why it does in the first place) but #839 tries to fix this issue.

**Comment by vuki656** [comment]
Oh okay, feel free to close this issue if it's not going to be used for tracking as I would consider my problem solved.

Thanks for the help.

**Comment by rklak** [comment]
It should goes like this in config, if you dont know how to setup it:
```
require('telescope').setup{
	defaults = {
		path_display={"smart"} 
	}
}
```

Check:
`:help telescope.setup()`

**Comment by nicos68** [comment]
> It should goes like this in config, if you dont know how to setup it:
> 
> ```
> require('telescope').setup{
> 	defaults = {
> 		path_display={"smart"} 
> 	}
> }
> ```
> 
> Check: `:help telescope.setup()`

Thanks, it worked with `find_files`! However I tried the setting with `Telescope lsp_references`, it doesn't seem to work.

**Comment by BlakeLeBlancGuernsey** [comment]
Similar results to @nicos68 

In the uses I've tested, it seems that some of the options ("smart", "shorten", etc) 

**Do work with**
- `builtin.buffers`
- `builtin.find_files`
- `builtin.live_grep`

**Do NOT work with**
- `builtin.oldfiles`
- `builtin.gitfiles`

Perhaps there is something I haven't configured properly?

This is what I'm currently rocking in my init.lua:

```
  -- telescope
  {
    'nvim-telescope/telescope.nvim',
    config = function()
      require("telescope").setup({
        defaults = {
          layout_config = {
            vertical = {
              width = 0.95
            },
            horizontal = {
              width = 0.95
            },
          },
          path_display = {
            "shorten"
          },
        },
      })
    end
  },
```

**Comment by searleser97** [comment]
> Similar results to @nicos68
> 
> In the uses I've tested, it seems that some of the options ("smart", "shorten", etc)
> 
> **Do work with**
> 
> * `builtin.buffers`
> * `builtin.find_files`
> * `builtin.live_grep`
> 
> **Do NOT work with**
> 
> * `builtin.oldfiles`
> * `builtin.gitfiles`
> 
> Perhaps there is something I haven't configured properly?
> 
> This is what I'm currently rocking in my init.lua:
> 
> ```
>   -- telescope
>   {
>     'nvim-telescope/telescope.nvim',
>     config = function()
>       require("telescope").setup({
>         defaults = {
>           layout_config = {
>             vertical = {
>               width = 0.95
>             },
>             horizontal = {
>               width = 0.95
>             },
>           },
>           path_display = {
>             "shorten"
>           },
>         },
>       })
>     end
>   },
> ```

Agree with this and also it doesn't work with extensions

**Comment by mturoci** [comment]
> Agree with this and also it doesn't work with extensions

Same here. Custom function for `path_display` works for `builtin.find_files`, but not for `builtin.git_status`.

**Comment by jamestrew** [comment]
I fixed the `path_display` option for `git_status` here https://github.com/nvim-telescope/telescope.nvim/pull/2881

I tested out `oldfiles` and `git_files`, but they appear to be working. Also looking at the code, they should work as well.

**Comment by sassanh** [comment]
@jamestrew is there a reason why `path_truncate` is not exposed in utils api?
Third parties and custom pickers would really benefit from it, specially considering it is using `calc_result_length` and it is too complicated and hacky for other developers to come to the conclusion they can achieve line length like that, I made it public for myself and now I can truncate paths in my custom picker like this:
```lua
utils.path_truncate(display_filename, 1, {})
```
I can't imagine how I could achieve this myself without reusing `path_truncate`
